app/slave/cluster_slave.py

```diff
@@ -173,18 +175,20 @@ def _do_build_teardown_and_reset(self, timeout=None):
         for executor in self.executors_by_id.values():
             executor.kill()
 
-        if not self._project_type:
-            return  # There is no build to tear down! (Slave is idle.)
+        # This lock keeps multiple teardown attempts from racing each other and throwing exceptions.
+        with self._build_teardown_lock:
```

Comment:
In my functional tests, I start a build and then tell slaves to shutdown.  There's a scenario where the build completes and a slave starts running teardown(), then it receives the shutdown command, tries running teardown() again, and throws a big error.  In this scenario I don't want an error: as long as teardown only runs once, then everything is working correctly.  The lock allows the second teardown() call to exit 'properly' because `if not self._project_type` is True.  

Some alternatives: 
1) Remove the lock and don't raise an error if the coin is already spent.
```
if not self._build_teardown_count.spend() or not self._project_type: #  Order matters!!
  return
```
2) Remove the lock and wrap every _do_build_teardown_and_reset with a try/catch and ignore the BuildTeardownError.

3) Keep the lock and remove the build_teardown_coin.

I'm leaning toward 1) I think